### ISSUE TITLE
gc-tags labels vs annotations 

### ISSUE THREAD

**Issue by mkmik** [kubecfg/kubecfg#159]
Initially we implement gc-tags in annotations but that makes it slow for us to find objects matching the tag (it requires a full scan of the apiserver), so we're migrating to using labels.

However @muxmuse found out that:

>  Some automatically created resources inherit labels of their "parents", but not annotations. E.g. Endpoints inherit labels (but not annotations) of the related Service and don't get an ownerReference, such that they will be garbage-collected on each kubecfg update when using tag-based garbage collection.

> Reproducible on docker-desktop v1.25.0 with

```
kubecfg update --gc-tag=test main.yaml
kubecfg update --gc-tag=test main.yaml
# INFO  Garbage collecting endpoints default.main (v1)
# main.yaml
---
apiVersion: apps/v1
kind: Deployment
metadata:
  name: main
spec:
  selector:
    matchLabels:
      app: main
  template:
    metadata:
      labels:
        app: main
    spec:
      containers:
      - name: main
        image: containous/whoami
---
apiVersion: v1
kind: Service
metadata:
  name: main
spec:
  ports:
  - port: 80
  selector:
    app: main
  type: ClusterIP
```

> This also applies to metrics.k8s.io/PodMetrics which inherit labels of the Pods they belong to, but get no ownerReference.

**Comment by camh-** [comment]
One solution to this would be to tag AND annotate resources, and only garbage collect those with both. The annotation would not necessarily need to be a gc-tag annotation but could just be a "created by kubecfg" annotation, so that kubecfg only garbage collected resources it has created itself.

**Comment by mkmik** [comment]
I wonder if we should just use ApplySet pruning. 
I found it hard to use ApplySet pruning but at least it would be "standard"

WDYT?

**Comment by camh-** [comment]
I didn't know about ApplySet pruning until now. The KEP (https://github.com/kubernetes/enhancements/issues/3659) seems to be in limbo in that there's a good stable alpha implementation but no-one seems to be driving it to graduate.

Personally I'd stick with kubecfg's current implementation with minor fixes - just based on the amount of time people seem to have for developing kubecfg (not a criticism). If ApplySet pruning graduates, it would make sense to adhere to that spec.

**Comment by camh-** [comment]
One current issue with tags vs annotations is it seems [gctag-migration2] has been partially (inadvertently?) implemented, which is what is causing this current garbage collecting endpoints issue - the comments in [`update.go`](https://github.com/kubecfg/kubecfg/blob/0468968aebc213a72703a8abdfdedbb433961731/pkg/kubecfg/update.go#L55) say:

> // NB: this is in phase1 of a migration from an annotation.
> // At this stage, both label+migration are written, but the
> // annotation (only) is still used to trigger GC. [gctag-migration]

But PR#150 (Use gc tags present in input) started using the label to determine if a resource should be garbage collected: https://github.com/kubecfg/kubecfg/pull/150/files#diff-ed0105b292e79171a755c9ee1408bca6121999464cac1f5f58bd0659dfe73f48R528-R532

> 	var tag string
>	tag = obj.GetLabels()[LabelGcTag]
>	if tag == "" {
>		tag = a[AnnotationGcTag]
>	}

I have a local build that replaces that with `tag := a[AnnotationsGcTag]` and I no longer have an issue with endpoints being garbage collected.

It could work that phase 2 of gctag-migration only starts using the labels to find the resources to check, and then checks those resources for the annotation. That would resolve this issue preventing kubecfg removing resources it should not.

**Comment by mkmik** [comment]
Yeah. Fwiw I almost always ended up using `kubecfg show | kubectl apply ` for various reasons (like server side apply support)

I wonder if it makes sense to keep supporting a parallel kubecfg update codepath or gradually make kubecfg update just a frontend to show + shelling out to kubectl apply (or using kubectl as library?)

That said I'm a strong supporter of backward compatibility so I'll do my best to not unnecessary break kubecfg update.

**Comment by mkmik** [comment]
(just to be clear: I'm all in for addressing this issue in the most immediate way)
